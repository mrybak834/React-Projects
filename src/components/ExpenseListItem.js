import React from 'react';
import { Link } from 'react-router-dom'

const ExpenseListItem =  ({ id, description, amount, timestamp }) => (
    <div>
        <h3><Link to={`edit/${id}`}>{description}</Link></h3>
        <p>${amount} - Created At: {timestamp}</p>
    </div>
);

export default ExpenseListItem;