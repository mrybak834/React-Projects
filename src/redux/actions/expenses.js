import uuid from 'uuid';


export const addExpense = ( 
    {
        description = '', 
        note = '', 
        amount = 0, 
        timestamp = 0
    } = {} 
) => (
    {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            description,
            note,
            amount,
            timestamp
        }
    }
);

export const removeExpense = (id = undefined) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    }
);

export const editExpense = (id = undefined, updates = {}) => (
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
);