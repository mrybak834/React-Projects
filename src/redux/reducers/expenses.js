const expensesReducerDefaultState = [];

export default ( state = expensesReducerDefaultState, action ) => {
    switch (action.type){
        case 'ADD_EXPENSE':{
            if (!action.expense){
                return state;
            }
            return state.concat(action.expense);
            // return [...state, action.expense];   // Slower than concat
        }
        case 'REMOVE_EXPENSE':{
            // No id defined, don't remove anything
            if (!action.id){
                return state;
            }

            return state.filter(({id}) => id !== action.id);
        }
        case 'EDIT_EXPENSE':{
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        // Overrides any of the expense properties above
                        ...action.updates
                    }
                }

                return expense;
            });
        }
        default:{
            return state;
        }
    }
};