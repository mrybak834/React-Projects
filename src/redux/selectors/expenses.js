import moment from 'moment';


export default (expenses, { text, sortBy, startDate, endDate }) => {

    return expenses.filter( (expense) => {
        // // Filter text
        // if ( text && ((expense.description !== text) && (expense.note !== text)) ){
        //     return false;
        // }
        // // Filter Dates
        // if ( (startDate && (expense.timestamp < startDate)) || (endDate && (expense.timestamp > endDate))){
        //     return false;
        // }

        const timestampMoment = moment(expense.timestamp);

        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = startDate ? startDate.isSameOrBefore(timestampMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(timestampMoment, 'day') : true;
        return textMatch && startDateMatch && endDateMatch;
    }).sort( (a,b) => {
        if (sortBy === 'date'){
            // -1 = a comes first
            return a.timestamp > b.timestamp ? -1 : 1;
        }
        // amount
        else {
            return a.amount > b.amount ? -1 : 1;
        }
    });
};