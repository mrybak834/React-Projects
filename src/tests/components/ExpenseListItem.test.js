import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';
import moment from 'moment';

test('Components: ExpenseListItem: Render Expense List Item', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);

    expect(wrapper).toMatchSnapshot();
});