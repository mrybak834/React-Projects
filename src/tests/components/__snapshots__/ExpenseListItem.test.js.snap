// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: ExpenseListItem: Render Expense List Item 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="edit/1"
    >
      test
    </Link>
  </h3>
  <p>
    $
    1
     - Created At: 
    864000000
  </p>
</div>
`;
